packages/commerce-sdk/src/core/loyalty/index.ts

```diff
@@ -0,0 +1,213 @@
+import { SDKOptions, ThirdwebSDK } from "@thirdweb-dev/sdk";
+import crypto from "crypto";
+import { NextApiRequest } from "next";
+import { LoyaltyRewardsParams, ResponseBody, RewardTokensParams } from "../../../types";
+import { GET_ORDER_BY_ID_QUERY } from "../../lib/queries";
+import { getRawBody, shopifyFetchAdminAPI } from '../../lib/utils';
+
+/**
+ * Gift ERC20 loyalty tokens to a customer using their wallet address.
+ * 
+ * @description This function is used to gift loyalty tokens to a customer using their wallet address. The customer must have purchased and item from your store in order to receive the loyalty tokens. The function will check the order ID and verify that the customer has purchased an item from your store. If the customer has purchased an item from your store, the function will gift the customer the loyalty tokens.
+ * 
+ * @example
+ * ```
+ * const txHash = await rewardPoints({
+    nextApiRequest: {{next_api_request}},
+    rewardAmount: 100,
+    tokenContractAddress: {{"contract_address"}},
+    webhookSecret: {{"webhook_secret"}},
+    chain: "goerli",
+    privateKey: {{"private_key"}},
+    shopifyAdminUrl: {{"shopify_admin_url"}},
+    shopifyAccessToken: {{"shopify_access_token"}},
+  });
+ * ```
+ * @returns Transaction hash
+ * @public
+ */
+
+export async function rewardTokensOnPurchase({
+  nextApiRequest,
+  webhookSecret,
+  tokenContractAddress,
+  gaslessRelayerUrl,
+  chain,
+  rewardAmount,
+  privateKey,
+  shopifyAdminUrl,
+  shopifyAccessToken,
+}: LoyaltyRewardsParams) {
+  const secretKey = webhookSecret;
+
+  const hmac = nextApiRequest.headers["x-shopify-hmac-sha256"];
+  const body = await getRawBody(nextApiRequest);
+  const hash = crypto
+    .createHmac("sha256", secretKey)
+    .update(body)
+    .digest("base64");
+
+    // Compare our hash to Shopify's hash
+    if (hash === hmac) {
+      const shopifyOrderId = nextApiRequest.headers["x-shopify-order-id"];
+
+      const response = await shopifyFetchAdminAPI({
+        shopifyAdminUrl,
+        shopifyAccessToken,
+        query: GET_ORDER_BY_ID_QUERY,
+        variables: {
+          id: `gid://shopify/Order/${shopifyOrderId}`,
+        },
+      });
+
+      const respBody = response.body as ResponseBody;
+
+      if (!respBody.data.order.lineItems) {
+        throw new Error("Order did not contain any line items");
+      }
+
+      const itemsPurchased = respBody.data.order.lineItems.edges.map(
+        (edge) => edge.node,
+      );
+
+      let wallet = "";
+      try {
+        wallet =
+          itemsPurchased[0].customAttributes.find(
+            (e: any) => e.key === "wallet",
+          )?.value || "";
+      } catch (e) {
+        throw new Error("No wallet address found in order's custom attributes");
+      }
```

Comment:
let's make *very* sure we document that this is how they have to attach the wallet address, also something to ask others (lazer / shopify) around to provide feedback if theres really no better way